### PR TITLE
refactor(clustering/rpc): tunings and fixes for rubustness

### PR DESCRIPTION
**Pull request by chronolaw**
### Summary

Please squash and merge.

KAG-5617

Related tickets:
KAG-5602

### Checklist

- [ ] The Pull Request has tests
- [ ] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

Fix #_[issue number]_

### CODE REVIEW

File: kong/db/strategies/off/init.lua [hunk 1]
```diff
@@ -176,7 +188,7 @@ local function select(self, pk, options)
   local ws_id = workspace_id(schema, options)
   local pk = pk_string(schema, pk)
   local key = item_key(schema.name, ws_id, pk)
-  return select_by_key(schema, key, false)
+  return select_by_key(schema, key, need_follow(ws_id))
```
Comment (chobits) [inline kong/db/strategies/off/init.lua:191]: if ws_id equals "*", i think we dont need to follow it. because you could select it with key directly

Comment (chronolaw) [reply kong/db/strategies/off/init.lua:191]: No, the global key stores idx_value, not the serialized value.

File: kong/db/declarative/import.lua [hunk 2]
```diff
@@ -289,13 +307,20 @@ local function _set_entity_for_txn(t, entity_name, item, options, is_delete)
...
     local cache_key = dao:cache_key(item)
-    local key = unique_field_key(entity_name, ws_id, "cache_key", cache_key)
 
-    -- store item_key or nil into lmdb
-    t:set(key, idx_value)
+    for _, wid in ipairs {ws_id, GLOBAL_WORKSPACE_TAG} do
+      local key = unique_field_key(entity_name, wid, "cache_key", cache_key)
+
+      -- store item_key or nil into lmdb
+      t:set(key, idx_value)
```
Comment (chobits) [inline kong/db/declarative/import.lua:322]: not a blocker, just a comment for other reviewers. TODO: we only want to save the entity with key containing specific ws_id, not a * ws_id, which does not consume too much memory space for lmdb. https://konghq.atlassian.net/browse/KAG-5704

File: kong/db/declarative/import.lua [hunk 3]
```diff
@@ -354,10 +383,18 @@ end
...
--- * <entity_name>|<ws_id>|<foreign_field_name>|<foreign_key>|<pk_string> -> <entity_name>|<ws_id>|*|<pk_string>
+--
+-- * <entity_name>|*|*|<pk_string> => serialized item
+-- * <entity_name>|<ws_id>|*|<pk_string> => actual item key
+--
+-- * <entity_name>|*|<unique_field_name>|sha256(field_value) => actual item key
+-- * <entity_name>|<ws_id>|<unique_field_name>|sha256(field_value) => actual item key
+--
+-- * <entity_name>|*|<foreign_field_name>|<foreign_key>|<pk_string> => actual item key
+-- * <entity_name>|<ws_id>|<foreign_field_name>|<foreign_key>|<pk_string> => actual item key
```
Comment (chobits) [inline kong/db/declarative/import.lua:397]: not a blocker, -> https://github.com/Kong/kong/pull/13771/files#r1821977274

**Review by chobits — APPROVED** [review]

File: kong/clustering/services/sync/rpc.lua [hunk 4]
```diff
@@ -261,13 +263,13 @@ local function do_sync()
...
       if old_entity and not wipe then
-        local res, err = delete_entity_for_txn(t, delta_type, old_entity, nil)
+        local res, err = delete_entity_for_txn(t, delta_type, old_entity, opts)
         if not res then
           return nil, err
         end
       end
 
-      local res, err = insert_entity_for_txn(t, delta_type, delta_row, nil)
+      local res, err = insert_entity_for_txn(t, delta_type, delta_row, opts)
```
[the remark this answers is not in the document]
Comment (chronolaw) [reply kong/clustering/services/sync/rpc.lua]: In `hooks.lua`, delta item will always has the field `ws_id`:
```
  local deltas = {
    {
      type = name,
      pk = pk,
      ws_id = ws_id,
      entity = is_delete and ngx_null or entity,
    },
  }
```

File: kong/db/declarative/export.lua [hunk 5]
```diff
@@ -359,7 +359,8 @@ local sync_emitter = {
 
   emit_entity = function(self, entity_name, entity_data)
     self.out_n = self.out_n + 1
-    self.out[self.out_n] = { type = entity_name , row = entity_data, version = self.sync_version, }
+    self.out[self.out_n] = { type = entity_name , row = entity_data, version = self.sync_version,
+                             ws_id = kong.default_workspace, }
```
[the remark this answers is not in the document]
Comment (chronolaw) [reply kong/db/declarative/export.lua:367]: Here `ws_id` is a fallback, if row(entity) has the field `ws_id`, `do_sync()` will use it but not delta.ws_id.

**Review by StarlightIbuki — APPROVED** [review]

**Review by bungle — APPROVED** [review]

**Comment by chronolaw** [comment]
Cherry-picked in https://github.com/Kong/kong-ee/pull/10460

**Pull request merged**